crates/schema-inference/src/server.rs

```diff
@@ -0,0 +1,277 @@
+                            } else {
+                                accumulator = Some(inferred_shape)
+                            }
+                            docs = docs + 1;
+                        }
+                    }
+                    Err(e) => return Err(Status::aborted(e.to_string())),
+                }
+            }
+            _ = owned_abort_channel.recv() => {
+                tracing::debug!("Aborting schma inference early! Processed {} documents", docs);
+                break
+            }
+        }
+    }
+
+    match accumulator {
+        Some(accum) => Ok(Some((accum, docs))),
+        None => Ok(None),
+    }
+}
+
+async fn reduce_shape_stream(
+    stream: impl Stream<Item = Result<Option<(Shape, u64)>, Status>>,
+) -> Result<Option<(Shape, u64)>, Status> {
+    let mut accumulator: Option<(Shape, u64)> = None;
+    tokio::pin!(stream);
+
+    while let Some(shape) = stream.next().await {
+        match shape {
+            Ok(Some((inferred_shape, docs_read))) => {
+                if let Some((accumulated_shape, docs_count)) = accumulator {
+                    accumulator = Some((
+                        shape::merge(accumulated_shape, inferred_shape),
+                        docs_count + docs_read,
+                    ))
+                } else {
+                    accumulator = Some((inferred_shape, docs_read))
+                }
+            }
+            Ok(None) => {}
+            Err(e) => return Err(e),
+        }
+    }
+
+    Ok(accumulator)
+}
+
+#[derive(Debug, clap::Args)]
+pub struct ServeArgs {
+    #[clap(long, value_parser, default_value_t = 50051)]
+    port: u16,
+    #[clap(long, value_parser, default_value = "[::1]")]
+    hostname: String,
+    /// URL for a Gazette broker that is a member of the cluster
+    #[clap(long, value_parser)]
+    broker_url: String,
+    /// Maximum number of seconds to run inference. This exists to preserve system performance and reliability in the face of large collections
+    #[clap(long, value_parser, default_value_t=10)]
+    inference_deadline: u32
+}
+
+impl ServeArgs {
+    #[tracing::instrument(skip(self))]
+    pub async fn run(&self) -> Result<(), anyhow::Error> {
+        let svc = InferenceServiceImpl {
+            broker_url: self.broker_url.clone(),
+            max_inference_secs: self.inference_deadline
+        };
+        let addr = format!("{}:{}", self.hostname, self.port).parse()?;
+
+        tracing::info!("🚀 Serving gRPC on {}", addr);
+
+        Server::builder()
+            .add_service(InferenceServiceServer::with_interceptor(svc, require_auth))
+            .serve(addr)
+            .await?;
+
+        Ok(())
+    }
+}
+
+struct Authentication {
+    token: String,
+}
+
+fn require_auth(mut req: Request<()>) -> Result<Request<()>, Status> {
+    match req.metadata().get("authorization").cloned() {
+        Some(t) /* TODO: validate JWT token */ => {
+            req.extensions_mut().insert(Authentication {
```

Comment:
Is the idea for this server to call the data-plane-gateway (which would enforce auth), or the other way around? Or is that still undecided? The passing along of the token makes it seem like this service would delegate to DPG for the reads, but then that sort of implies that we have a separate ingress for this service, which also kind of doesn't make sense. I vaguely recall there being some conversation on this topic, so apologies if I've forgotten.

Reply:
Welll... I am also not quite sure. This token is passed to `connect_journal_client`, which ultimately is used to set an `AuthHeader` interceptor on the `JournalClient` that gets created, so I assumed it was needed.

Reply:
I'm thinking it would be better to have data-plane-gateway handle authz and delegate to schema-inference, which would connect to gazette directly (and thus not need the authorization header). It certainly won't hurt for the authorization header to be present, but I don't think it should be _required_.

That being said, IDK how this all squares with the potentially changing view of this being framed as a runtime concern, similar to how stats are handled. We should discuss that soon, but I don't think it should block merging this PR.